### PR TITLE
[RFR] Review Resource component

### PR DESCRIPTION
**Pull request by djhi**
- [x] Review Resource component
- [x] Updated tests
- [x] Updated documentation

Also included the Loading component from #1299 by @tkvw (Thanks :))

### CODE REVIEW

File: packages/ra-example/posts.js [hunk 1]
```diff
@@ -153,27 +153,16 @@ export const PostCreate = ({ ...props }) => (
         <SimpleForm
             toolbar={<PostCreateToolbar />}
             defaultValue={{ average_note: 0 }}
-            validate={values => {
```
Comment (fzaninotto) [inline packages/ra-example/posts.js]: not good, please use another view for debugging field-level validation, we need this one to test form-level validation

File: packages/react-admin/src/AdminRoutes.js [hunk 2]
```diff
@@ -17,16 +15,11 @@ export class AdminRoutes extends Component {
     // Can't use null or undefined here as authClient may return any those values
     state = { permissions: initialPermissions };
```
Comment (fzaninotto) [inline packages/react-admin/src/AdminRoutes.js]: I'd set the `childrenToRender` key to `[]` to make it explicit.

File: packages/react-admin/src/AdminRoutes.js [hunk 3]
```diff
@@ -40,98 +33,115 @@ export class AdminRoutes extends Component {
...
+        if (
+            typeof nextProps.children === 'function' &&
+            nextState.permissions !== this.state.permissions
+        ) {
+            const { children } = nextProps;
+
+            const childrenFuncResult = children(nextState.permissions);
+            if (childrenFuncResult.then) {
+                childrenFuncResult.then(resolvedChildren => {
+                    this.setState({
```
Comment (fzaninotto) [inline packages/react-admin/src/AdminRoutes.js]: Is it a good practice to update the state (`childrenToRender`) while it was just updated (`premissions`)? Won't it cause a double render?

Comment (djhi) [reply packages/react-admin/src/AdminRoutes.js]: Yes but it only renders a Loading screen if there are no children to render anyway. Are you worried about updated permissions ? In this case, yes too but I don't see any other way to do it

File: packages/ra-example/posts.js [hunk 4]
```diff
@@ -172,7 +183,6 @@ const emptyKeycode = [];
 // Store the validation functions array in a variable to avoid
```
Comment (fzaninotto) [inline packages/ra-example/posts.js]: please remove the comment, too

**Pull request merged**